### ISSUE TITLE
No audio on macos with jack

### ISSUE THREAD

**Issue by fdch** [pure-data/pure-data#1384]
Well, these are two cases where there is no audio with `jackd -d coreaudio -r 48000` (jack version 1.9.19 installed from homebrew on Mojave 10.14.6 ). 

- Using Miller's Pd-0.51-4, I get `Can't open Jack (it seems not to be installed)` on the Pd Console.
- Using a compiled [1] pd from latest git. Pd connects to Jack (after I turn on DSP I see the pure_data jack ports) [2], but there is no dsp processing on the Test Audio and Midi patch. 

Both cases are either with connecting to Jack from the `Media > Jack` menu or running pd like `./path/to/bin/pd -jack -channels 2 -r 48000`. In both cases, portaudio works.

[1] configured like this:
```
./configure --enable-fftw --disable-jack-framework --enable-jack
```
[2] to my surprise, I noticed that pd jack ports now start at 1, not 0 as before...

**Comment by fdch** [comment]
>  does Pd still react otherwise? 
Yes, it does.
> are `callbacks` turned on or off?
When `callbacks` are turned off, it does not work. Otherwise, there is audio!

**Comment by fdch** [comment]
One more thing. As [this comment](https://github.com/pure-data/pure-data/issues/1190#issuecomment-790190424) says on the issue you mention, when starting with `brew services start jack`, pd works.

However, callbacks act funny here:
- callbacks off: normal sine tone on the Test Audio and Midi patch
- callbacks on: sine tone shifts pitch down, almost a half-step

**Comment by umlaeute** [comment]
i'm not sure i can make sense of the last two comments.
it seems that:
- *everything*¹ works if you start jack with `brew services start jack`
- if you don't start jack, it works with `callbacks: ON` but not otherwise (i have no idea why you would get sound if JACK is not running)

> callbacks on: sine tone shifts pitch down, almost a half-step

that looks a lot like a 44.1kHz vs 48kHz issue

**Comment by fdch** [comment]
> - if you don't start jack, it works with `callbacks: ON` but not otherwise (i have no idea why you would get sound if JACK is not running)
> 

No, that is not the case. When I start jack with `jackd -d coreaudio -r 48000` and then pd with `pd -jack -channels 2` I get audio when `callbacks` are ON, but no audio when `callbacks` are OFF. 


> that looks a lot like a 44.1kHz vs 48kHz issue

yes, but I did not change the sample rate :)

**Comment by umlaeute** [comment]
you say:

> but no audio when `callbacks` are OFF.

and you also say:

> callbacks off: normal sine tone on the Test Audio and Midi patch

so how do you get the "normal sine tone" if there is "no audio"?

---

and regarding the pitch shift:
> yes, but I did not change the sample rate :)

- so what sr is your soundcard running at? (i could imagine JACK to try setting 48kHz and *believing* it succeeded, when in reality it does not)
- in your last comment you don't specify the samplerate when starting Pd; does it help if you do set it to 48kHz?
- what happens if you try running jackd/pd at 44kHz?

**Comment by fdch** [comment]
> so how do you get the "normal sine tone" if there is "no audio"?

So, the callbacks acting funny comment was after starting jack with: `brew services start jack`. I never used that command, so I am not sure what that does to samplerate. Pd was probably trying to use 48kHz, because it probably remembered my preferences (I mostly use 48). Jack was probably going at 44.1. I will confirm this when I get to my computee.

**Comment by fdch** [comment]
Indeed, when I started jack with `brew services start jack`, jack was running at 44.1kHz, which would explain pitch shift.

**Comment by umlaeute** [comment]
that's at least something.

otoh, Pd should get the correct samplerate from JACK and force it's internal rate to that value.
Which it does here (running `cf44c8bd4e0e9825a92d2d67ccc78d451f6bb21f` on Debian/GNU linux (unstable) with `jack2_1.9.19~dfsg-2`; JACK is running at 44.1kHz, Pd attempts to use 48kHz):

![Pd-jack48](https://user-images.githubusercontent.com/214034/135575114-e01c22ea-0ff2-446e-85cc-a832a2aff37a.png)

As you can see, it detects that JACK is running at 44.1kHz and uses that (so the pitch is fine), ignoring the value passed from the cmdline; in the Audio Preferences the actual samplerate is displayed and cannot set 48kHz in the audio preferences.

With Pd-0.51-4 (as packaged in Debian), behaviour is slightly different:
- the sr-value passed on cmdline is ignored and the one suggested by JACK is used (so the pitch is OK)
- opening the audio preferences shows me the requested (but unused) rate (`48000`), and if i click on <kbd>Apply</kbd> (after changing nothing) Pd starts using this samplerate (and the pitch goes down).

so do you (only?) get the wrong pitch with Pd-0.51-4 and using the audio preferences *somehow* (eg. for selecting JACK as the backend)?

**Comment by umlaeute** [comment]
actually the samplerate *displayed* in the preferences (of Pd-git cf44c8bd4e0e9825a92d2d67ccc78d451f6bb21f) is *not* the one reported by JACK:
Like this:

![which rate is it?](https://user-images.githubusercontent.com/214034/135578161-6edbfa56-c3ce-4443-b04b-1df9c7edf070.png)

this is indeed confusing.

The pitch is OK though. so Pd is apparently running at 44.1kHz as dictated by JACK.
This can also be confirmed with `[samplerate~]`.

**Comment by fdch** [comment]
TL;DR Pd Audio dialog does not report correct samplerate with jack. Audio does not work with `-noprefs` and callbacks `off`. 

Here are some tests, all with built-in input and output on macos.  I hope they are useful and clear, please let me know otherwise.  I can keep going at it if you send me more instructions.  

**Where is the .pdsettings file located on macos?**

Here's pd and jack versions I have:

```
fd:~ fd$ jackd --version
jackdmp version 1.9.19 tmpdir /tmp protocol 9
fd:~ fd$ ls -la $(which jackd)
lrwxr-xr-x  1 fd  admin  31 Sep 28 21:10 /usr/local/bin/jackd -> ../Cellar/jack/1.9.19/bin/jackd
fd:~ fd$ pd -version
Pd-0.51.4 ("") compiled 21:19:11 Sep 28 2021
```
NOTE: I have `pd` as an alias to the app built by `make app` 


## with -noprefs

1. Start jack with `jackd -d coreaudio -r 44100`
2. Start pd with `pd -noprefs -jack`
3. Pd Audio Dialog reports samplerate at 48000
4. Open `Media > Test Audio and Midi` patch
   1. Dsp turns `ON`
   2. Input NOT Reported on channels 1 and 2
   3. Turn testtone `ON`
5. `callbacks` are `OFF`: No audio in Pd
6. Toggle callbacks `ON`, click 'Apply':
   1. Input Reported on channels 1 and 2
   2. Testttone sounds at lower pitch
7. Toggle `callbacks` `OFF`, click 'Apply': No audio in Pd

![Screen Shot 2021-10-01 at 6 16 54 PM](https://user-images.githubusercontent.com/13335792/135687144-1792f3e0-f73d-44ee-b7c6-e36ba4aac6fc.png)


## without -noprefs

1. Start jack with `jackd -d coreaudio -r 44100`
2. Start pd with `pd -jack`
3. Pd Audio Dialog reports samplerate at 44100
4. Open `Media > Test Audio and Midi`
   1. Dsp turns `ON`
   2. Input Reported on channels 1 and 2
   3. Turn sinetone `ON`
5. `callbacks` are `OFF`: testtone sounds at 440 Hz
6. Toggle callbacks `ON`, click 'Apply': Audio breaks up for a short while, then no change in pitch.
7. Toggle `callbacks` `OFF`, click 'Apply': testtone sounds at 440 Hz
![Screen Shot 2021-10-01 at 5 50 54 PM](https://user-images.githubusercontent.com/13335792/135686841-a924ac95-446e-43c7-a8c5-b1104fcb76d2.png)

## without -noprefs, but specifying -r in pd

1. Start jack with `jackd -d coreaudio -r 96000`
2. Start pd with `pd -jack -r 96000`
3. Pd Audio Dialog reports samplerate at 44100 (!)
4. Open `Media > Test Audio and Midi`
   1. Dsp turns `ON`
   2. Input Reported on channels 1 and 2
   3. Turn sinetone `ON`
5. `callbacks` are `OFF`: testtone sounds at 440 Hz
6. Toggle callbacks `ON`, click 'Apply': Audio breaks up for a short while, then pitch shifts about minor 9th up.
7. Toggle `callbacks` `OFF`, click 'Apply': testtone remains a minor 9th up
8. Toggle DSP `OFF` and `ON`: testtone sounds at 440 Hz

![Screen Shot 2021-10-01 at 6 09 06 PM](https://user-images.githubusercontent.com/13335792/135686876-2aca33b2-d127-4ce1-89e3-7d3cbcf402fe.png)

**Comment by umlaeute** [comment]
that lools like your samplerate is actually changeable in the audio preferences - which in turn strongly indicates that you are *not* running the latest Pd from git (e.g. `711a28fd370ec6edce4b0c3029040379250a5076`). (unfortunately the "-version" doesn't give any hint about which Pd version you are running)


> Pd Audio dialog does not report correct samplerate with jack.

just what i was saying :-)

> Audio does not work with `-noprefs` and callbacks *off*.

so that is really a logical `&&`, right? "Audio does not work if both `-noprefs` is a cmdline flag **and** callbacks are turned off".
(as opposed to: "Audio does not work with either `-noprefs` or 'no callbacks'")

This is of course a bit weird, as `-noprefs` should actually only provide defaults...

anyhow, does `pd -noprefs -callback -jack` work then?


> Where is the `.pdsettings? file located on macOS?

nowhere. on macOS the settings are stored in a `.plist` file (like with any well-behaving macOS application)
The key is `org.puredata.pd`, if i'm not mistaken.

**Comment by fdch** [comment]
So, I have [this commit](https://github.com/fdch/pure-data/commit/cf44c8bd4e0e9825a92d2d67ccc78d451f6bb21f) in my fork which I thought was the latest. Should I go for the one you mention?

**Comment by umlaeute** [comment]
> Should I go for the one you mention?


No, this should be sufficient.
Just make sure that you can indeed create a listbox gatom with your version of Pd...

**Comment by fdch** [comment]
I can, yes

**Comment by danomatika** [comment]
>on macOS the settings are stored in a .plist file (like with any well-behaving macOS application)
>The key is org.puredata.pd, if i'm not mistaken.

Well documented in the platform-specific readme...

https://github.com/pure-data/pure-data/blob/master/mac/README.txt#L147

**Comment by danomatika** [comment]
WHat's the status on this? Working now or...?

**Comment by fdch** [comment]
If I run `pd -noprefs -callback -jack`, the audio works, and the samplerate matches Jack's once audio is on. But, if I run `pd -noprefs -jack`, there does not seem to be any audio being processed. 

If I run `pd -noprefs -pa`, the audio is set to 48k and works well. Using `-callback` here makes it choppy. Fiddling with callback on the Audio Settings dialog makes Pd freeze and I have to kill it.
